base/dates/io.jl

```diff
@@ -1,29 +1,97 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
-# TODO: optimize this
-function Base.string(dt::DateTime)
-    y,m,d = yearmonthday(days(dt))
-    h,mi,s = hour(dt),minute(dt),second(dt)
-    yy = y < 0 ? @sprintf("%05i",y) : lpad(y,4,"0")
-    mm = lpad(m,2,"0")
-    dd = lpad(d,2,"0")
-    hh = lpad(h,2,"0")
-    mii = lpad(mi,2,"0")
-    ss = lpad(s,2,"0")
-    ms = millisecond(dt) == 0 ? "" : string(millisecond(dt)/1000.0)[2:end]
-    return "$yy-$mm-$(dd)T$hh:$mii:$ss$(ms)"
-end
-Base.show(io::IO,x::DateTime) = print(io,string(x))
-function Base.string(dt::Date)
-    y,m,d = yearmonthday(value(dt))
-    yy = y < 0 ? @sprintf("%05i",y) : lpad(y,4,"0")
-    mm = lpad(m,2,"0")
-    dd = lpad(d,2,"0")
-    return "$yy-$mm-$dd"
-end
-Base.show(io::IO,x::Date) = print(io,string(x))
-
-### Parsing
+"""
+A type of token in a date time string
+
+each subtype must define
+
+    tryparsenext(t::TokenType, str, i, [locale])
+
+and
+
+    format(io, t::TokenType, [locale])
+"""
+abstract AbstractDateToken
+
+"""
+Locale type for dates. By default `DateLocale{:english}()`
+is used. This object is passed as the last argument to
+`tryparsenext` and `format` defined for each `AbstractDateToken` type.
+"""
+immutable DateLocale{lang} end
```

Comment:
I strongly recommend against baking `lang` into a type parameter. This will actually make the API more awkward to extend than presenting it to the user as a dict.